cpg-core/src/main/java/de/fraunhofer/aisec/cpg/graph/Extensions.kt

```diff
@@ -136,6 +136,104 @@ class StatementNotFound : Exception()
 
 class DeclarationNotFound(message: String) : Exception(message)
 
+fun Node.followPrevDFGEdgesUntilHit(predicate: (Node) -> Boolean): List<Node> {
+    val result = mutableListOf<Node>()
+    val alreadySeen = mutableListOf<Node>()
+    val worklist = mutableListOf<Node>()
+    worklist.add(this)
+
+    while (worklist.isNotEmpty()) {
+        val currentNode = worklist.removeFirst()
+        alreadySeen.add(currentNode)
+        for (prev in currentNode.prevDFG) {
+            if (predicate(prev)) {
+                result.add(prev)
+            }
+            if (!alreadySeen.contains(prev)) {
+                worklist.add(prev)
+            }
+        }
+    }
+
+    return result
+}
+
+/**
+ * Returns a list of nodes which are data flow paths between the starting node [this] and the end
+ * node fulfilling [predicate]. Paths which do not end at such a node are not included in the
+ * result. Hence, if the return value is a non-empty list, a data flow from [this] to such a node is
+ * **possible but not mandatory**.
+ */
+fun Node.followNextDFGEdgesUntilHit(predicate: (Node) -> Boolean): List<List<Node>> {
+    // Looks complicated but at least it's not recursive...
+    // result: List of paths (between from and to)
+    val result = mutableListOf<List<Node>>()
+    // The list of paths where we're not done yet.
+    val worklist = mutableListOf<List<Node>>()
+    worklist.add(listOf(this)) // We start only with the "from" node (=this)
+
+    while (worklist.isNotEmpty()) {
+        val currentPath = worklist.removeFirst()
+        // The last node of the path is where we continue. We get all of its outgoing DFG edges and
+        // follow them
+        for (next in currentPath.last().nextDFG) {
+            // Copy the path for each outgoing DFG edge and add the next node
+            val nextPath = mutableListOf<Node>()
+            nextPath.addAll(currentPath)
+            nextPath.add(next)
+            if (predicate(next)) {
+                // We ended up in the node "to", so we're done. Add the path to the results.
+                result.add(nextPath)
+            }
+            // The next node is new in the current path (i.e., there's no loop), so we add the path
+            // with the next step to the worklist.
+            if (!currentPath.contains(next)) {
+                worklist.add(nextPath)
+            }
+        }
+    }
+
+    return result
+}
+
+/**
+ * Returns a list of nodes which are evaluation paths between the starting node [this] and the end
+ * node fulfilling [predicate]. Paths which do not end at such a node are not included in the
+ * result. Hence, if the return value is a non-empty list, a path from [this] to such a node is
+ * **possible but not mandatory**.
+ */
+fun Node.followNextEOGEdgesUntilHit(predicate: (Node) -> Boolean): List<List<Node>> {
```

Comment:
What's the particular difference between the "new" (until hit) and "old" functions without the suffix?

Reply:
The main difference is that the "new" functions return all paths satisfying `predicate` whereas the "old" functions return only one of them.